File: packages/python/elizaos/features/basic_capabilities/providers/current_time.py
from __future__ import annotations

from datetime import UTC, datetime
from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_provider_spec
from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_provider_spec("CURRENT_TIME")


async def get_current_time_context(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    _ = runtime, message, state

    now = datetime.now(UTC)

    iso_timestamp = now.isoformat()
    human_readable = now.strftime("%A, %B %d, %Y at %H:%M:%S UTC")
    date_only = now.strftime("%Y-%m-%d")
    time_only = now.strftime("%H:%M:%S")
    day_of_week = now.strftime("%A")
    unix_timestamp = int(now.timestamp())

    context_text = f"# Current Time\n- Date: {date_only}\n- Time: {time_only} UTC\n- Day: {day_of_week}\n- Full: {human_readable}\n- ISO: {iso_timestamp}"

    return ProviderResult(
        text=context_text,
        values={
            "currentTime": iso_timestamp,
            "currentDate": date_only,
            "dayOfWeek": day_of_week,
            "unixTimestamp": unix_timestamp,
        },
        data={
            "iso": iso_timestamp,
            "date": date_only,
            "time": time_only,
            "dayOfWeek": day_of_week,
            "humanReadable": human_readable,
            "unixTimestamp": unix_timestamp,
        },
    )


current_time_provider = Provider(
    name=_spec["name"],
    description=_spec["description"],
    get=get_current_time_context,
    dynamic=_spec.get("dynamic", True),
)

File: packages/typescript/src/features/basic-capabilities/providers/currentTime.ts
import { requireProviderSpec } from "../../../generated/spec-helpers.ts";
import type {
	IAgentRuntime,
	Memory,
	Provider,
	State,
} from "../../../types/index.ts";

// Get text content from centralized specs
const spec = requireProviderSpec("CURRENT_TIME");

/**
 * Current time provider function that retrieves the current date and time
 * in various formats for use in time-based operations or responses.
 *
 * @param _runtime - The runtime environment of the bot agent.
 * @param _message - The memory object containing message data.
 * @returns An object containing the current date and time data in various formats.
 */
export const currentTimeProvider: Provider = {
	name: spec.name,
	description: spec.description,
	dynamic: spec.dynamic ?? true,
	get: async (_runtime: IAgentRuntime, _message: Memory, _state: State) => {
		const now = new Date();

		const isoTimestamp = now.toISOString();
		const unixTimestamp = Math.floor(now.getTime() / 1000);

		const options = {
			timeZone: "UTC",
			dateStyle: "full" as const,
			timeStyle: "long" as const,
		};
		const humanReadable = new Intl.DateTimeFormat("en-US", options).format(now);

		const dateOnly = now.toISOString().split("T")[0];
		const timeOnly = now.toISOString().split("T")[1].split(".")[0];
		const dayOfWeek = now.toLocaleDateString("en-US", {
			weekday: "long",
			timeZone: "UTC",
		});

		const contextText = `# Current Time
- Date: ${dateOnly}
- Time: ${timeOnly} UTC
- Day: ${dayOfWeek}
- Full: ${humanReadable}
- ISO: ${isoTimestamp}`;

		return {
			text: contextText,
			values: {
				currentTime: isoTimestamp,
				currentDate: dateOnly,
				dayOfWeek: dayOfWeek,
				unixTimestamp: unixTimestamp,
			},
			data: {
				iso: isoTimestamp,
				date: dateOnly,
				time: timeOnly,
				dayOfWeek: dayOfWeek,
				humanReadable: humanReadable,
				unixTimestamp: unixTimestamp,
			},
		};
	},
};
